### PR TITLE
[Keymap] Add BB8520 trackpad support for CrKbd

### PR DESCRIPTION
**Pull request by vlukash**
Add BlackBerry 8520 trackpad support for CrKbd via two new keymaps.

## Description

I've created an adapter that allows adding the BlackBerry 8520 trackpad mod to the CrKbd via additional 'adapter' PCB. The adapter PCB uses Elite-C board as there are not enough pins on ProMicro.

The trackpad add-on requires different firmware for the left and right keyboards as 'adapter PCB' is used only on the right board and it rewires some pins that were originally set for ProMicro (like hardware SPI pins).

Please see more details here:
- https://github.com/vlukash/corne-trackpad
- https://vlukash.com/2019/01/15/trackpad-in-keycap-corne-crkbd-keyboard

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+#include "bootloader.h"
+#ifdef PROTOCOL_LUFA
+  #include "lufa.h"
+  #include "split_util.h"
+#endif
+#ifdef SSD1306OLED
+  #include "ssd1306.h"
+#endif
+
+extern keymap_config_t keymap_config;
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c]:
```suggestion
```

File: keyboards/crkbd/keymaps/vlukash_trackpad_right/keymap.c [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+#include "report.h"
+
+#ifdef PROTOCOL_LUFA
+  #include "lufa.h"
+  #include "split_util.h"
+#endif
+
+extern bool isScrollMode;
+
+extern keymap_config_t keymap_config;
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/vlukash_trackpad_right/keymap.c]:
```suggestion
```

File: keyboards/crkbd/keymaps/vlukash_trackpad_right/keymap.c [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+  RAISE,
+  ADJUST,
+  BACKLIT,
+  RGBRST,
+  MBTN1,
+  SCRL
+};
+
+enum macro_keycodes {
+  KC_SAMPLEMACRO,
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/vlukash_trackpad_right/keymap.c]: This enum is not used, and is a holdover from the old macro system. It should be removed.

File: keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+  RAISE,
+  ADJUST,
+  BACKLIT,
+  RGBRST,
+  MBTN1,
+  SCRL
+};
+
+enum macro_keycodes {
+  KC_SAMPLEMACRO,
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c]: This enum is not used, and is a holdover from the old macro system. It should be removed.

File: keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+extern uint8_t is_master;
+
+// Each layer gets a name for readability, which is then used in the keymap matrix below.
+// The underscores don't mean anything - you can have a layer called STUFF or any other name.
+// Layer names don't all need to be of the same length, obviously, and you can also skip them
+// entirely and just use numbers.
+#define _QWERTY 0
+#define _LOWER 3
+#define _RAISE 4
+#define _ADJUST 16
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c]: Non-consecutive layer numbers will generate extra empty layers in between, unnecessarily increasing flash usage. The better way to define layers is with an enum, eliminating the need to number them manually:

```c
enum layer_names {
    _QWERTY,
    _LOWER,
    _RAISE,
    _ADJUST
};
```

File: keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+#define KC_CTLA CTL_T(KC_A)
+#define KC_CTLSC CTL_T(KC_SCLN)
+#define KC_SFTZ SFT_T(KC_Z)
+#define KC_SFTSL SFT_T(KC_SLSH)
+
+#define KC_MBTN1  MBTN1
+#define KC_SCRL   SCRL
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+  [_QWERTY] = LAYOUT_kc( \
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c]: The backslashes at the end of these lines are not necessary. You can keep them if you want as it isn't hurting anything, but IMO it looks neater and less "hack-y".

File: keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+    // set_timelog();
+  }
+
+  switch (keycode) {
+    case QWERTY:
+      if (record->event.pressed) {
+        persistent_default_layer_set(1UL<<_QWERTY);
+      }
+      return false;
+      break;
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c]: `break` after `return` is dead code.

File: keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+      if (record->event.pressed) {
+        layer_on(_RAISE);
+        update_tri_layer_RGB(_LOWER, _RAISE, _ADJUST);
+      } else {
+        layer_off(_RAISE);
+        update_tri_layer_RGB(_LOWER, _RAISE, _ADJUST);
+      }
+      return false;
+      break;
+    case ADJUST:
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c]: I think this keycode can be replaced by a simple `MO(_ADJUST)`.

File: keyboards/crkbd/keymaps/vlukash_trackpad_left/rules.mk [hunk 9]
```diff
@@ -0,0 +1,33 @@
+
+# Build Options
+#   change to "no" to disable the options, or define them in the Makefile in
+#   the appropriate keymap folder that will get included automatically
+#
+BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration(+1000)
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/vlukash_trackpad_left/rules.mk]: You only need to specify the options that differ from the base rules.mk here, so most of these can go.

File: keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c (hunk 1 above)
Comment (vlukash) [reply keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c]: Removed, thanks!

File: keyboards/crkbd/keymaps/vlukash_trackpad_right/keymap.c (hunk 2 above)
Comment (vlukash) [reply keyboards/crkbd/keymaps/vlukash_trackpad_right/keymap.c]: Removed, thanks!

File: keyboards/crkbd/keymaps/vlukash_trackpad_right/keymap.c (hunk 3 above)
Comment (vlukash) [reply keyboards/crkbd/keymaps/vlukash_trackpad_right/keymap.c]: Cleaned up!

File: keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c (hunk 4 above)
Comment (vlukash) [reply keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c]: Cleaned up!

File: keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c (hunk 5 above)
Comment (vlukash) [reply keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c]: Got it, updated with enum. Thanks for details!

File: keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c (hunk 6 above)
Comment (vlukash) [reply keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c]: Removed the backslashes.

File: keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c (hunk 8 above)
Comment (vlukash) [reply keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c]: Removed ADJUST keycode, it wasn't used in this keymap.

File: keyboards/crkbd/keymaps/vlukash_trackpad_left/rules.mk (hunk 9 above)
Comment (vlukash) [reply keyboards/crkbd/keymaps/vlukash_trackpad_left/rules.mk]: Fixed both rules.mk, thanks!

File: keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c (hunk 7 above)
Comment (vlukash) [reply keyboards/crkbd/keymaps/vlukash_trackpad_left/keymap.c]: Fixed.

**Comment by drashna** [comment]
Travis CI errors are unrelated to this change.  

Thanks!

File: keyboards/crkbd/rev1/matrix.c [hunk 10]
```diff
@@ -93,6 +93,44 @@ uint8_t matrix_cols(void)
     return MATRIX_COLS;
 }
 
+void tx_rx_leds_init(void)
```
Comment (drashna) [inline keyboards/crkbd/rev1/matrix.c]: @foostan Any issues with these changes?

Comment (vlukash) [reply keyboards/crkbd/rev1/matrix.c]: @foostan, @drashna, this is just a way to get access to PD5 and PB0 AtMega32u4 pins. On the regular ProMicro they're hardwired to two TX and RX LEDs, but on Elite-C they're available as general purpose pins, and in my case, I need them to do a matrix scan. 

I believe CrKbd uses them just for visual monitoring of the communication process between two halves. 
Please let me know if there's a better way to do this.

Comment (foostan) [reply keyboards/crkbd/rev1/matrix.c]: I checked it, and no problem.

Comment (vlukash) [reply keyboards/crkbd/rev1/matrix.c]: @foostan thanks for verifying this!

**Pull request merged**